tests/unit/test_utils.py

```diff
@@ -332,26 +425,62 @@ def confirm_files(self):
                 "mode: %s, expected mode: %s" % (mode, expected_mode)
             )
 
-    def test_unpack_tgz(self, data):
+    def test_unpack_tar(self):
         """
-        Test unpacking a *.tgz, and setting execute permissions
+        Test unpacking a *.tar, and setting execute permissions
         """
-        test_file = data.packages.join("test_tar.tgz")
+        test_file = self.create_tar_files()
         untar_file(test_file, self.tempdir)
         self.confirm_files()
         # Check the timestamp of an extracted file
         file_txt_path = os.path.join(self.tempdir, 'file.txt')
         mtime = time.gmtime(os.stat(file_txt_path).st_mtime)
-        assert mtime[0:6] == (2013, 8, 16, 5, 13, 37), mtime
+        assert mtime[0:6] == self.file_mod_time, mtime
 
-    def test_unpack_zip(self, data):
+    def test_unpack_zip(self):
         """
         Test unpacking a *.zip, and setting execute permissions
         """
-        test_file = data.packages.join("test_zip.zip")
+        test_file = self.create_zip_files()
         unzip_file(test_file, self.tempdir)
         self.confirm_files()
 
+    def test_unpack_tar_fail(self):
+        """
+        Test unpacking a *.tar which contains .. path, and expect exception
+        """
+        test_file = self.create_tar_files()
+        dir_name = os.path.dirname(test_file)
+        default_cwd = os.getcwd()
+        os.chdir(dir_name)
+        with tarfile.open(test_file, 'a') as tar:
+            upper_level_file = '../upper_level_file.txt'
+            with open(upper_level_file, 'w+') as fh:
+                fh.write('upper level file')
+            tar.add(upper_level_file)
+        os.chdir(default_cwd)
+        with pytest.raises(InstallationError,
+                           match=r'.*is trying to install outside location.*'):
+            untar_file(test_file, self.tempdir)
+
+    def test_unpack_zip_fail(self):
+        """
+        Test unpacking a *.zip which contains .. path, and expect exception
+        """
+        test_file = self.create_zip_files()
+        dir_name = os.path.dirname(test_file)
+        default_cwd = os.getcwd()
+        os.chdir(dir_name)
+        with zipfile.ZipFile(test_file, 'a') as zip:
+            upper_level_file = '../upper_level_file.txt'
+            with open(upper_level_file, 'w+') as fh:
+                fh.write('upper level file')
+            zip.write(upper_level_file)
+        os.chdir(default_cwd)
```

Comment:
This test can be made a lot simpler. For example, if you define a helper function `create_zip_file(dest_path, files)` that creates a zip file with the relative paths given to it, you could replace all of the above lines with a single line like this:

```python
create_zip_file(zip_path, files=['temp.txt', '../outside.txt'])
```

The current test uses a setup that is a lot more complicated than needed (and that uses a zip file with many more files than needed) to test the error.

You should also use the `tmpdir` fixture which will simplify things further.

Reply:
I am fixing the original tests as they are using a tar/zip file and like u said, people are difficult to review what's inside the file. I dont even know why the original PR get merged. So i had followed the approach to provide a tar/zip file to create a test, but this got rejected. Then I decide to review all the tests including the old, all tests are creating the zip/tar files at run time and it does introduce complexity. 

So what i see the problem now is the communication is really low efficient, i had spent my whole day re-writing the tests cases  now it turns out it is not what as expected

Reply:
With the approach I’ve been trying to suggest to you, the PR should be relatively simple. I never meant to suggest that other tests should be refactored as part of this PR. You’re always welcome to ask questions in advance if you’re unsure of what’s being asked.

Reply:
ok, thanks. for new untar and unzip failure test cases, i would simply create a zip/tar file that contains a txt file and ../outside.txt file like you suggest then. Original tests will remain unchanged

Reply:
btw, is it possible i can run these travis CI tests on my local mac. waiting for CI tests to complete and report failure really takes time

Reply:
You need to install `tox`, and then you should be able to run, for example:

```
$ tox -e py36 tests/functional/test_install.py::test_install_conflict_results_in_warning
```

You may need to change the Python version, and this shows you how to include the individual test file and/or test function name.

Reply:
> i would simply create a zip/tar file that contains a txt file and ../outside.txt file like you suggest then. Original tests will remain unchanged

Yes. The contents of the file can be any fake string data. Also, you should be able to create the zip files without needing to create the original files on the file system (just write directly to the ZipFile object I'm pretty sure).